types/bent/index.d.ts

```diff
@@ -0,0 +1,41 @@
+// Type definitions for bent 7.0
+// Project: https://github.com/mikeal/bent#readme
+// Definitions by: Ovyerus <https://github.com/me>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+// TypeScript Version: 2.7
+
+/// <reference types="node" />
+
+import { PassThrough, Stream } from 'stream';
+
+type HttpMethod = 'GET' | 'POST' | 'DELETE' | 'PUT' | 'PATCH' | 'HEAD' | 'OPTIONS' | 'CONNECT' | 'TRACE';
+type StatusCode = number;
+type BaseUrl = string;
+interface Headers { [x: string]: any; }
+
+type Json = {[x: string]: any} | any[];
+type ValidResponse = bent.BentResponse | string | Buffer | ArrayBuffer | Json;
+
+declare function bent(...args: bent.Options[]): bent.RequestFunction;
+
+// Can't use these until we get some sort of pattern checking for strings, ala https://github.com/microsoft/TypeScript/issues/6579
+// Otherwise, `string` from BaseUrl just eats up the extra string literals here, meaning it ends up defaulting to the first type.
+// declare function bent(...args: bent.Options[]): bent.RequestFunction<bent.BentResponse>;
```

Comment:
These should work fine. Write this instead:
```ts
declare function bent(type: 'string', ...args: bent.Options[]): bent.RequestFunction<string>;
declare function bent(type: 'buffer', ...args: bent.Options[]): bent.RequestFunction<Buffer | ArrayBuffer>;
declare function bent(type: 'json', ...args: bent.Options[]): bent.RequestFunction<Json>;
declare function bent(...args: bent.Options[]): bent.RequestFunction<bent.BentResponse>;
```

Order matters, and being more specific helps.

Reply:
The problem is bent allows the type - or any other argument - at any
position in the parameters list, so it would break for some people's usage.

Reply:
I think it's OK to only give the more specific types for certain forms - if anything, it's a little better than what's there right now.